### PR TITLE
Make all model lookups go through the container

### PR DESCRIPTION
**Pull request by fivetanley**
move factories off constructor and delegate to the container for lookups

This commit changes the way factories are located by Ember Data.
Previously, factories were stuffed onto the constructor of a
DS.Model subclass.

This was problematic for a few reasons:

- Classes for relationships defined via hasMany(Object) (instead of
  hasMany('string-name') were different than what was in the container.
  This caused several issues for Ember CLI and Globals Mode users alike.
- Inconsistency with the rest of the way lookups work in Ember.
- Zalgo

Now, relationship information can be accessed off the "type" in the
relationships "meta" hash. If you need to lookup a factory in your app,
you need to use this.store.modelFor, or this.store.modelFactoryFor if
you prefer to check that the model exists without throwing an exception.
This change makes all model factory lookups go through the container.
This should hopefully kill `Ember.MODEL_FACTORY_INJECTIONS = true;`.

### CODE REVIEW

**Comment by fivetanley** [comment]
cc @igorT @mixonic @stefanpenner

File: packages/ember-data/lib/serializers/json-serializer.js [hunk 1]
```diff
@@ -158,13 +158,14 @@ export default Serializer.extend({
   */
   normalize: function(type, hash) {
     if (!hash) { return hash; }
+    var model = this.store.modelFor(type);
```
Comment (fivetanley) [inline packages/ember-data/lib/serializers/json-serializer.js]: not sure at this point if serializers should receive type or typeKey.

File: packages/ember-data/lib/system/relationships/state/belongs-to.js [hunk 2]
```diff
@@ -60,14 +60,18 @@ BelongsToRelationship.prototype._super$addRecord = Relationship.prototype.addRec
...
+  var modelClass = this.store.modelFor(type);
+  Ember.assert("You cannot add a '" + newRecord.constructor.typeKey + "' record to the '" + this.record.constructor.typeKey + "." + this.key +"'. " + "You can only add a '" + modelClass.typeKey + "' record to this relationship.", (function () {
+    if (modelClass.__isMixin) {
+      return modelClass.__mixin.detect(newRecord);
     }
-    if (Ember.MODEL_FACTORY_INJECTIONS) {
-      type = type.superclass;
+
+    // TODO: There seems to be a weird cylic error with loader js here.
+    if (modelClass.superclass && modelClass.superclass !== DS.Model) {
```
Comment (fivetanley) [inline packages/ember-data/lib/system/relationships/state/belongs-to.js]: Need to talk to stef here. This inheritance check is wacky. Not sure what to check for wrt injections.

File: packages/ember-data/lib/system/relationships/state/has-many.js [hunk 3]
```diff
@@ -84,12 +85,12 @@ ManyRelationship.prototype.removeRecordFromOwn = function(record, idx) {
...
 ManyRelationship.prototype.notifyRecordRelationshipAdded = function(record, idx) {
-  var type = this.relationshipMeta.type;
-  Ember.assert("You cannot add '" + record.constructor.typeKey + "' records to the " + this.record.constructor.typeKey + "." + this.key + " relationship (only '" + this.belongsToType.typeKey + "' allowed)", (function () {
+  var type = this.store.modelFor(this.relationshipMeta.type);
+  Ember.assert("You cannot add '" + record.constructor.typeKey + "' records to the " + this.record.constructor.typeKey + "." + this.key + " relationship (only '" + this.belongsToType + "' allowed)", (function () {
     if (type.__isMixin) {
       return type.__mixin.detect(record);
     }
-    if (Ember.MODEL_FACTORY_INJECTIONS) {
+    if (type.superclass && type.superclass !== DS.Model) {
```
Comment (fivetanley) [inline packages/ember-data/lib/system/relationships/state/has-many.js]: wacky inheritance check here too.

File: packages/ember-data/lib/system/store.js [hunk 4]
```diff
@@ -1421,12 +1421,14 @@ Store = Service.extend({
   */
   modelFor: function(key) {
     var factory;
+    var singularizedKey;
 
     if (typeof key === 'string') {
-      factory = this.modelFactoryFor(key);
+      singularizedKey = singularize(key);
```
Comment (fivetanley) [inline packages/ember-data/lib/system/store.js]: this can go away since modelFactoryFor normalizes.

File: packages/ember-data/lib/system/relationships/state/belongs-to.js (hunk 2 above)
Comment (stefanpenner) [inline packages/ember-data/lib/system/relationships/state/belongs-to.js]: i believe once we have this commit inplace, we can begin to remove then `extend` in the container, which breaks alot of inheritance checks, This that i believe we can clean ^^ up.

**Comment by stefanpenner** [comment]
This looks great.

so to confirm, no more instances (like store) now exist on the factories?

**Comment by fivetanley** [comment]
@bmac I'm removing this from the milestone. It needs some battletesting before we bring it in.

File: packages/ember-data/lib/serializers/json-serializer.js (hunk 1 above)
Comment (igorT) [inline packages/ember-data/lib/serializers/json-serializer.js]: While we are at it, can we maybe figure out a better name than model for the actual class. Something like using type & typeKey or typeClass. Model implies an actual record, very easy to get confused

File: packages/ember-data/lib/adapters/rest-adapter.js [hunk 5]
```diff
@@ -502,7 +502,7 @@ export default Adapter.extend(BuildURLMixin, {
   */
   findBelongsTo: function(store, record, url, relationship) {
     var id   = get(record, 'id');
-    var type = record.constructor.typeKey;
+    var type = store.modelFor(record.constructor.typeKey);
```
Comment (stefanpenner) [inline packages/ember-data/lib/adapters/rest-adapter.js]: `record.constructor` is the type

File: packages/ember-data/lib/system/relationships/ext.js [hunk 6]
```diff
@@ -277,9 +278,9 @@ Model.reopenClass({
       var possibleRelationships = relationshipsSoFar || [];
 
       var relationshipMap = get(inverseType, 'relationships');
-      if (!relationshipMap) { return; }
+      if (!relationshipMap) { return possibleRelationships; }
 
-      var relationships = relationshipMap.get(type);
+      var relationships = relationshipMap.get(type.typeKey);
```
Comment (stefanpenner) [inline packages/ember-data/lib/system/relationships/ext.js]: we should try and make `type` always the string and if its something else call it the factory/modelConstructor or something

File: packages/ember-data/lib/system/relationship-meta.js [hunk 7]
```diff
@@ -8,7 +8,7 @@ export function typeForRelationshipMeta(store, meta) {
     if (meta.kind === 'hasMany') {
       typeKey = singularize(typeKey);
     }
-    type = store.modelFor(typeKey);
```
Comment (stefanpenner) [inline packages/ember-data/lib/system/relationship-meta.js]: `store` is now no longer needed as an argument to this function. which means  `relatedTypesDescriptor`  and `relationshipsByNameDescriptor` should no longer need the store either.

File: packages/ember-data/lib/system/many-array.js [hunk 8]
```diff
@@ -226,7 +226,9 @@ export default Ember.Object.extend(Ember.MutableArray, Ember.Evented, {
     var type = get(this, 'type');
     var record;
 
-    Ember.assert("You cannot add '" + type.typeKey + "' records to this polymorphic relationship.", !get(this, 'isPolymorphic'));
+    var klass = store.modelFactoryFor(type);
+
+    Ember.assert("You cannot add '" + klass.typeKey + "' records to this polymorphic relationship.", !get(this, 'isPolymorphic'));
```
Comment (kball) [inline packages/ember-data/lib/system/many-array.js]: Do we know that in this situation modelFactoryFor will never return null?  Because if not, this will break...

File: packages/ember-data/lib/system/store.js [hunk 9]
```diff
@@ -1445,8 +1447,9 @@ Store = Service.extend({
   },
 
   modelFactoryFor: function(key) {
-    if (this.container.has('model:' + key)) {
-      return this.container.lookupFactory('model:' + key);
+    var singularized = singularize(key);
```
Comment (kball) [inline packages/ember-data/lib/system/store.js]: In some places we're using singularized, and other places fully normalizing with this._normalizeTypeKey.  Shouldn't we normalize the same way throughout?

File: packages/ember-data/lib/system/many-array.js (hunk 8 above)
Comment (fivetanley) [inline packages/ember-data/lib/system/many-array.js]: Yeah, modelFactoryFor won't raise an exception. modelFor will.

Comment (fivetanley) [inline packages/ember-data/lib/system/many-array.js]: Oh wait, I read that wrong. You sir, are correct!

File: packages/ember-data/lib/system/store.js (hunk 9 above)
Comment (fivetanley) [inline packages/ember-data/lib/system/store.js]: That is a good point! looking into it now. Thanks for taking the time to review! Makes me more confident that we'll catch more things before this gets merged in.

File: packages/ember-data/lib/system/relationships/ext.js [hunk 10]
```diff
@@ -76,8 +76,8 @@ var relationshipsByNameDescriptor = Ember.computed(function() {
   this.eachComputedProperty(function(name, meta) {
     if (meta.isRelationship) {
       meta.key = name;
-      var relationship = relationshipFromMeta(this.store, meta);
-      relationship.type = typeForRelationshipMeta(this.store, meta);
+      var relationship = relationshipFromMeta(meta);
+      relationship.type = typeForRelationshipMeta(meta);
```
Comment (kball) [inline packages/ember-data/lib/system/relationships/ext.js]: I believe this is extraneous.  Looking at relationshipFromMeta it appears to already define type using typeForRelationshipMeta

File: packages/ember-data/lib/system/relationships/ext.js [hunk 11]
```diff
@@ -207,6 +207,7 @@ Model.reopenClass({
     @method inverseFor
     @static
     @param {String} name the name of the relationship
+    @param {DS.Store} store the current store
```
Comment (kball) [inline packages/ember-data/lib/system/relationships/ext.js]: I don't think this comment belongs here...

File: packages/ember-data/lib/system/relationships/state/has-many.js [hunk 12]
```diff
@@ -84,12 +85,12 @@ ManyRelationship.prototype.removeRecordFromOwn = function(record, idx) {
 };
 
 ManyRelationship.prototype.notifyRecordRelationshipAdded = function(record, idx) {
-  var type = this.relationshipMeta.type;
-  Ember.assert("You cannot add '" + record.constructor.typeKey + "' records to the " + this.record.constructor.typeKey + "." + this.key + " relationship (only '" + this.belongsToType.typeKey + "' allowed)", (function () {
+  var type = this.store.modelFor(this.relationshipMeta.type);
```
Comment (kball) [inline packages/ember-data/lib/system/relationships/state/has-many.js]: Should this also use typeForRelationshipMeta instead of trying to directly access the type?

File: packages/ember-data/lib/system/relationships/state/belongs-to.js (hunk 2 above)
Comment (kball) [inline packages/ember-data/lib/system/relationships/state/belongs-to.js]: I'm new to Ember Data so I don't really understand the backstory here, but is the situation of multiple layers of inheritance supported?  And if so, does that cause issues?  Because this appears to only explicitly handle one layer (both before and now)

E.g.
```
var Person = DS.Model.extend(/*..define some relationship..*/)
var Employee = Person.extend(/*...*/)
var Contractor = Employee.extend(/*...*/)
```

**Comment by fivetanley** [comment]
ping @igorT @stefanpenner

File: packages/ember-data/lib/system/relationships/state/belongs-to.js (hunk 2 above)
Comment (fivetanley) [inline packages/ember-data/lib/system/relationships/state/belongs-to.js]: Hey, sorry for the late reply. Yah, that use case should be and is currently supported.

**Comment by fivetanley** [comment]
you can try this out by using the following in your bower.json file:

```javascript
{
  "ember-data": "components/ember-data#always-lookup-models-through-container"
}
```

**Comment by fivetanley** [comment]
@stefanpenner i see you are reading your email :P

**Comment by stefanpenner** [comment]
i'll do one pass now, then another later :)

**Comment by stefanpenner** [comment]
got distracted by es7 decorators, will look later :P

File: packages/ember-data/lib/serializers/rest-serializer.js [hunk 13]
```diff
@@ -282,7 +282,7 @@ var RESTSerializer = JSONSerializer.extend({
 
       // legacy support for singular resources
       if (isPrimary && Ember.typeOf(value) !== "array" ) {
-        primaryRecord = this.normalize(primaryType, value, prop);
+        primaryRecord = this.normalize(type, value, prop);
```
Comment (igorT) [inline packages/ember-data/lib/serializers/rest-serializer.js]: Why this change? Shouldn't this be primary type?

File: packages/ember-data/lib/system/relationship-meta.js [hunk 14]
```diff
@@ -1,14 +1,14 @@
...
   var typeKey, type;
 
   typeKey = meta.type || meta.key;
   if (typeof typeKey === 'string') {
     if (meta.kind === 'hasMany') {
-      typeKey = singularize(typeKey);
+      typeKey = normalizeTypeKey(typeKey);
     }
-    type = store.modelFor(typeKey);
+    type = typeKey;
```
Comment (igorT) [inline packages/ember-data/lib/system/relationship-meta.js]: what is going on here, didn't we standardize the naming?

File: packages/ember-data/lib/serializers/rest-serializer.js (hunk 13 above)
Comment (fivetanley) [inline packages/ember-data/lib/serializers/rest-serializer.js]: i think so, yeah. i'll change it back.

File: packages/ember-data/lib/system/relationships/ext.js [hunk 15]
```diff
@@ -179,7 +178,7 @@ Model.reopenClass({
   */
   typeForRelationship: function(name) {
     var relationship = get(this, 'relationshipsByName').get(name);
-    return relationship && relationship.type;
+    return relationship && this.store.modelFor(relationship.type);
```
Comment (stefanpenner) [inline packages/ember-data/lib/system/relationships/ext.js]: this `store` shouldn't exist, as the goal of this work is to remove instance state from the factories. `this` here I believe is the factory.

File: packages/ember-data/lib/system/normalize-type-key.js [hunk 16]
```diff
@@ -0,0 +1,17 @@
+import {singularize} from 'ember-inflector/lib/system/string';
+
+var camelize = Ember.String.camelize;
```
Comment (fivetanley) [inline packages/ember-data/lib/system/normalize-type-key.js]: not sure why this came up in the diff.

File: packages/ember-data/lib/system/many-array.js [hunk 17]
```diff
@@ -230,7 +230,13 @@ export default Ember.Object.extend(Ember.MutableArray, Ember.Evented, {
     var type = get(this, 'type');
     var record;
 
-    Ember.assert("You cannot add '" + type.typeKey + "' records to this polymorphic relationship.", !get(this, 'isPolymorphic'));
+    var klass = store.modelFactoryFor(type);
```
Comment (mixonic) [inline packages/ember-data/lib/system/many-array.js]: It is pretty spooky that what `type` contains is unknown this far into the code. Shouldn't this normalization step happen earlier, like on `init`?

File: packages/ember-data/lib/system/relationships/ext.js [hunk 18]
```diff
@@ -273,13 +272,13 @@ Model.reopenClass({
       inverseKind = possibleRelationships[0].kind;
     }
 
-    function findPossibleInverses(type, inverseType, relationshipsSoFar) {
+    function findPossibleInverses(modelClass, inverseType, relationshipsSoFar) {
```
Comment (mixonic) [inline packages/ember-data/lib/system/relationships/ext.js]: the factory was called `typeClass` earlier, and here is `modelClass`. Should probably pick one.
